### PR TITLE
Add neural search default processor for non OpenAI/Cohere scenario

### PR DESCRIPTION
**Pull request by zane-neo**
### Description
ml-commons has two default pre/post process function which are for OpenAI and Cohere and written in painless script. There's no default pre/post process function for neural search plugin text embedding case, if user want to use neural search with remote model to text embedding, user has to write complex painless script which is a heavy burden. This PR do two things:

1. Add default pre/post process function for neural search plugin text embedding and once user defines their model output to List<List<Float>> as suggested, then in the connector pre/post process can be defined to strings like: `connector.pre_process.neural_search.text_embedding` or `connector.post_process.neural_search.text_embedding`, which is pretty easy for user to use neural search text embedding.
2. Change default OpenAI & Cohere painless script to java code to make them more readable to both coder and user.
 
### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/1304
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jngz-es** [comment]
I have a high level comment. From my understanding, the processors of connector is only about the format transformation between ml-commons and llms like OpenAI, Cohere etc. About the format transform part, I don't think neural search is a dependency for ml-commons, the only dependency should be on what llm is using. In other words, what format transforming looks like is only dependent on the llm we are using. We may have default pre/post processors, but not specific to neural search, it should be generic.

File: common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java [hunk 1]
```diff
@@ -5,61 +5,64 @@
...
-                "        if (i < embeddings.length - 1) {\n" +
-                "          builder.append(\",\");\n" +
-                "        }\n" +
-                "      }\n" +
-                "      builder.append(\"]\");\n" +
-                "      \n" +
-                "      return builder.toString();\n    ");
+        JSON_PATH_EXPRESSION.put(OPENAI_EMBEDDING, "$.data[*].embedding");
+        JSON_PATH_EXPRESSION.put(COHERE_EMBEDDING, "$.embeddings");
+        JSON_PATH_EXPRESSION.put(NEURAL_SEARCH_EMBEDDING, "$[*]");
```
Comment (jngz-es) [inline common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java:32]: Not sure for what model we expect the result is List<List<Float>>? How does the customer use this model?

File: common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java [hunk 2]
```diff
@@ -6,44 +6,37 @@
...
-
-        PRE_PROCESS_FUNCTIONS.put(TEXT_DOCS_TO_OPENAI_EMBEDDING_INPUT, "\n    StringBuilder builder = new StringBuilder();\n" +
-                        "    builder.append(\"\\\"\");\n" +
-                        "    builder.append(params.text_docs[0]);\n" +
-                        "    builder.append(\"\\\"\");\n" +
-                        "    def parameters = \"{\" +\"\\\"input\\\":\" + builder + \"}\";\n" +
-                        "    return  \"{\" +\"\\\"parameters\\\":\" + parameters + \"}\";");
+        PRE_PROCESS_FUNCTIONS.put(TEXT_DOCS_TO_COHERE_EMBEDDING_INPUT, cohereTextEmbeddingPreProcess());
+        PRE_PROCESS_FUNCTIONS.put(TEXT_DOCS_TO_OPENAI_EMBEDDING_INPUT, openAiTextEmbeddingPreProcess());
+        PRE_PROCESS_FUNCTIONS.put(NEURAL_SEARCH_EMBEDDING_INPUT, openAiTextEmbeddingPreProcess());
```
Comment (jngz-es) [inline common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java:32]: Do we only need a default pre-processor, not neural search?

**Comment by zane-neo** [comment]
> I have a high level comment. From my understanding, the processors of connector is only about the format transformation between ml-commons and llms like OpenAI, Cohere etc. About the format transform part, I don't think neural search is a dependency for ml-commons, the only dependency should be on what llm is using. In other words, what format transforming looks like is only dependent on the llm we are using. We may have default pre/post processors, but not specific to neural search, it should be generic.

The processor could be for non llms, e.g. our pre-trained model. Neural search is not a dependency for ml-commons, here the neural search might not a good name since it sounds more generic, but the main purpose here is to give user a default approach to transform the inputs from neural search plugin, e.g. TextEmbeddingProcessor to the expected format that user deployed pre-trained model in thirdparty platform, e.g. SageMaker.

File: common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java (hunk 1 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java:32]: For text embedding model, including OpenAI/Cohere embedding and our pre-trained text embedding models.

File: common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java (hunk 2 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java:32]: This is for neural search plugin text embedding use case, we can fine-tune the name though.

File: common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java [hunk 3]
```diff
@@ -6,44 +6,37 @@
...
+import java.util.function.Function;
 
 public class MLPreProcessFunction {
 
-    private static Map<String, String> PRE_PROCESS_FUNCTIONS;
+    private static final Map<String, Function<List<String>, Map<String, Object>>> PRE_PROCESS_FUNCTIONS = new HashMap<>();
     public static final String TEXT_DOCS_TO_COHERE_EMBEDDING_INPUT = "connector.pre_process.cohere.embedding";
     public static final String TEXT_DOCS_TO_OPENAI_EMBEDDING_INPUT = "connector.pre_process.openai.embedding";
 
+    public static final String NEURAL_SEARCH_EMBEDDING_INPUT = "connector.pre_process.neural_search.text_embedding";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java:19]: `connector.pre_process.default`.

May be we can add more comment to describe how are we making this default pre-processing function. But I assume people can easily assume if no other pre-processing works, it should fall back to `default` one.

**Review by jngz-es — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java [hunk 4]
```diff
@@ -5,61 +5,64 @@
...
 import java.util.Map;
+import java.util.function.Function;
 
 public class MLPostProcessFunction {
 
-    private static Map<String, String> POST_PROCESS_FUNCTIONS;
     public static final String COHERE_EMBEDDING = "connector.post_process.cohere.embedding";
     public static final String OPENAI_EMBEDDING = "connector.post_process.openai.embedding";
 
+    public static final String NEURAL_SEARCH_EMBEDDING = "connector.post_process.default";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java:22]: IS this default post process function will only be for embedding? If yes, then may be we can say: `connector.post_process.default.embedding`?

and the static variable name can be: `DEFAULT_EMBEDDING` (to match other two?)

File: common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java [hunk 5]
```diff
@@ -5,61 +5,64 @@
...
-                "                 \"\\\"data_type\\\":\\\"\" + dataType + \"\\\",\" +\n" +
-                "                 \"\\\"shape\\\":\" + shape + \",\" +\n" +
-                "                 \"\\\"data\\\":\" + params.data[0].embedding +\n" +
-                "                 \"}\";\n" +
-                "      return json;\n    ");
+    public static Function<List<List<Float>>, List<ModelTensor>> buildModelTensorList() {
+        return numbersList -> {
+            List<ModelTensor> modelTensors = new ArrayList<>();
+            if (numbersList == null) {
+                throw new IllegalArgumentException("NumbersList is null when applying build-in post process function!");
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java:42]: what do we mean by numberlist? Is that the list of embeddings? `numbers` is bit confusing here.

File: common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java [hunk 6]
```diff
@@ -6,44 +6,37 @@
...
+import java.util.function.Function;
 
 public class MLPreProcessFunction {
 
-    private static Map<String, String> PRE_PROCESS_FUNCTIONS;
+    private static final Map<String, Function<List<String>, Map<String, Object>>> PRE_PROCESS_FUNCTIONS = new HashMap<>();
     public static final String TEXT_DOCS_TO_COHERE_EMBEDDING_INPUT = "connector.pre_process.cohere.embedding";
     public static final String TEXT_DOCS_TO_OPENAI_EMBEDDING_INPUT = "connector.pre_process.openai.embedding";
 
+    public static final String NEURAL_SEARCH_EMBEDDING_INPUT = "connector.pre_process.default";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java:19]: same comment here (like post_process) about renaming for pre_process default.

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 7]
```diff
@@ -48,13 +43,13 @@ public static String toUTF8(String rawString) {
         return utf8EncodedString;
     }
 
-    public static Map<String, Object> fromJson(String jsonStr, String defaultKey) {
+    public static Map<String, Object> fromJson(String input, String defaultKey) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:46]: any reason to change the variable name? may be we can rename to: `inputJsonStr`?

File: common/src/main/java/org/opensearch/ml/common/utils/GsonUtil.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package org.opensearch.ml.common.utils;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.stream.JsonReader;
+
+public class GsonUtil {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/GsonUtil.java:7]: Not quite sure what are we exactly achieving with extending this class? May be we can add more comments to describe importance of this class?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java [hunk 9]
```diff
@@ -54,91 +58,112 @@ public static RemoteInferenceInputDataSet processInput(MLInput mlInput, Connecto
...
         return inputData;
     }
+    private static RemoteInferenceInputDataSet processTextDocsInput(TextDocsInputDataSet inputDataSet, Connector connector, Map<String, String> parameters, ScriptService scriptService) {
+        List<String> docs = new ArrayList<>(inputDataSet.getDocs());
+        Optional<ConnectorAction> predictAction = connector.findPredictAction();
+        if (predictAction.isEmpty()) {
+            throw new IllegalArgumentException("no predict action found");
+        }
+        String preProcessFunction = predictAction.get().getPreProcessFunction();
+        if (preProcessFunction == null) {
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:90]: I assume end user doesn't need to provide this `preProcessFunction`? By default it will be attached to the predictAction?

**Comment by dhrubo-os** [comment]
Are we planning to merge this PR in main?

File: common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java (hunk 5 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/connector/MLPostProcessFunction.java:42]: Changed the name to embeddings.

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 7 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java]: I was going to support non json format response, so changed the parameter name for better representation. Then found we don't have such case after discussion, removed the code in function body but missed this parameter name, changed back to jsonStr.

File: common/src/main/java/org/opensearch/ml/common/utils/GsonUtil.java (hunk 8 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/utils/GsonUtil.java:7]: I think this is common pattern to create a util class instead of `new` the target class everywhere, the beneficial are:
1. Less memory usage since only have one Gson instance in the GsonUtil.
2. Better encapsulation and better code organization since all Gson related code are in one place.
3. Cleaner code: no need to create Gson instance everywhere.

Comment (dhrubo-os) [reply common/src/main/java/org/opensearch/ml/common/utils/GsonUtil.java:7]: I'm not quite sure about the beneficial pointed out here:

1. We are already doing the same [here](https://github.com/opensearch-project/ml-commons/blob/2.x/common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java#L27-L29). I don't think we improved anything specially here. 
2. In all the `fromJson` methods we are just invoking the `fromJson` method from `Gson`. So I don't think we are achieving better code organization here rather than adding extra code. 
3. Same as 1

**Review by jngz-es — DISMISSED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java (hunk 9 above)
Comment (zane-neo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java]: Good finding, this can be attached to the predictAction by default.

File: common/src/main/java/org/opensearch/ml/common/utils/GsonUtil.java (hunk 8 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/utils/GsonUtil.java:7]: I missed this one, but I find another case [here](https://github.com/opensearch-project/ml-commons/blob/ff0d73460ee03e7f550c9383529a6627f4ac7ed3/ml-algorithms/src/main/java/org/opensearch/ml/engine/ModelHelper.java#L55). I think if we all follow the principle to use the Gson in StringUtils, then this GsonUtil is no more needed, I'll change this back but let's pay attention to this in the future code review.

**Comment by zane-neo** [comment]
> Are we planning to merge this PR in main branch first?

Yes, this will be merged to main branch and then backport to 2.x branch.

**Review by jngz-es — DISMISSED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java [hunk 10]
```diff
@@ -54,91 +58,110 @@ public static RemoteInferenceInputDataSet processInput(MLInput mlInput, Connecto
...
+            throw new IllegalArgumentException("no predict action found");
+        }
+        String preProcessFunction = predictAction.get().getPreProcessFunction();
+        preProcessFunction = preProcessFunction == null ? MLPreProcessFunction.DEFAULT_EMBEDDING_INPUT : preProcessFunction;
+        if (MLPreProcessFunction.contains(preProcessFunction)) {
+            Map<String, Object> buildInFunctionResult = MLPreProcessFunction.get(preProcessFunction).apply(docs);
+            return RemoteInferenceInputDataSet.builder().parameters(convertScriptStringToJsonString(buildInFunctionResult)).build();
+        } else {
+            if (preProcessFunction.contains("${parameters")) {
+                StringSubstitutor substitutor = new StringSubstitutor(parameters, "${parameters.", "}");
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:96]: Looks like we are using this `${parameters.` in multiple places, what do you think about using a Static variable for this?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ScriptUtils.java [hunk 11]
```diff
@@ -5,52 +5,36 @@
 
 package org.opensearch.ml.engine.utils;
 
-import com.google.gson.Gson;
-import org.opensearch.ml.common.connector.MLPostProcessFunction;
-import org.opensearch.ml.common.connector.MLPreProcessFunction;
-import org.opensearch.ml.common.utils.StringUtils;
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ScriptUtils.java:11]: Not sure why do we have to remove it from here and then use that in the code?

File: common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java [hunk 12]
```diff
@@ -6,44 +6,37 @@
...
+import java.util.function.Function;
 
 public class MLPreProcessFunction {
 
-    private static Map<String, String> PRE_PROCESS_FUNCTIONS;
+    private static final Map<String, Function<List<String>, Map<String, Object>>> PRE_PROCESS_FUNCTIONS = new HashMap<>();
     public static final String TEXT_DOCS_TO_COHERE_EMBEDDING_INPUT = "connector.pre_process.cohere.embedding";
     public static final String TEXT_DOCS_TO_OPENAI_EMBEDDING_INPUT = "connector.pre_process.openai.embedding";
 
+    public static final String DEFAULT_EMBEDDING_INPUT = "connector.pre_process.default.embedding";
```
[the remark this answers is not in the document]
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java]: done

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java (hunk 10 above)
Comment (zane-neo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:96]: this is existing code just moved to different place, beside this there are many other places using this string, we should make it a static variable but maybe in next PR

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ScriptUtils.java (hunk 11 above)
Comment (zane-neo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ScriptUtils.java:11]: Because the Pre/Post process function has nothing to do with painless script, they are java functions, ScriptUtils' single responsibility is to execute customized painless script.

Comment (dhrubo-os) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ScriptUtils.java:11]: I mean we removed import : `import org.opensearch.ml.common.utils.StringUtils;`

then used in the code like: `org.opensearch.ml.common.utils.StringUtils.fromJson(resultJson, "result");`

any advantage on using like this?

Comment (zane-neo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ScriptUtils.java:11]: Oh, this one. I think it's because I was using apache common StringUtils in this class as well so make this a fully qualified name, but apache one was removed, so we can use import StringUtils now.

**Review by dhrubo-os — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]
